tests/test_FLEbasis2d.py

```diff
@@ -0,0 +1,274 @@
+from aspire.volume import Volume
+
+from ._basis_util import UniversalBasisMixin
+
+DATA_DIR = os.path.join(os.path.dirname(__file__), "saved_test_data")
+
+
+class FLEBasis2DTestCase(TestCase, UniversalBasisMixin):
+    L = 8
+    dtype = np.float32
+
+    def setUp(self):
+        self.basis = FLEBasis2D((self.L, self.L), dtype=self.dtype)
+
+    # even and odd images with all guaranteed epsilons from paper
+    @parameterized.expand(
+        [
+            [32, 1e-4],
+            [32, 1e-7],
+            [32, 1e-10],
+            [32, 1e-14],
+            [33, 1e-4],
+            [33, 1e-7],
+            [33, 1e-10],
+            [33, 1e-14],
+        ]
+    )
+    # check closeness guarantees for fast vs dense matrix method
+    def testFastVDense_T(self, L, epsilon):
+        basis = FLEBasis2D(L, epsilon=epsilon, dtype=np.float64)
+        dense_b = basis.create_dense_matrix()
+
+        # create sample particle
+        x = self.create_images(L, 1).asnumpy()
+        xvec = x.reshape((L**2, 1))
+
+        # explicit matrix multiplication
+        result_dense = dense_b.T @ xvec
+        # fast evaluate_t
+        result_fast = basis.evaluate_t(Image(x))
+
+        relerr = self.relerr(result_dense.T, result_fast)
+        self.assertTrue(relerr < epsilon)
+
+    # even and odd images with all guaranteed epsilons from paper
+    @parameterized.expand(
+        [
+            [32, 1e-4],
+            [32, 1e-7],
+            [32, 1e-10],
+            [32, 1e-14],
+            [33, 1e-4],
+            [33, 1e-7],
+            [33, 1e-10],
+            [33, 1e-14],
+        ]
+    )
+    def testFastVDense(self, L, epsilon):
+        basis = FLEBasis2D(L, epsilon=epsilon, dtype=np.float64)
+        dense_b = basis.create_dense_matrix()
+
+        # get sample coefficients
+        x = self.create_images(L, 1)
+        # hold input test data constant (would depend on epsilon parameter)
+        coeffs = FLEBasis2D(L, epsilon=1e-4, dtype=np.float64).evaluate_t(x)
+
+        result_dense = dense_b @ coeffs.T
+        result_fast = basis.evaluate(coeffs).asnumpy()
+
+        relerr = self.relerr(result_dense, result_fast)
+        self.assertTrue(relerr < epsilon)
+
+    @parameterized.expand(
+        [
+            [32, 1e-4],
+            [32, 1e-7],
+            [32, 1e-10],
+            [32, 1e-14],
+            [33, 1e-4],
+            [33, 1e-7],
+            [33, 1e-10],
+            [33, 1e-14],
+        ]
+    )
+    def testEvaluateExpand(self, L, epsilon):
+        # compare result of evaluate() vs more accurate expand()
+        basis = FLEBasis2D(L, epsilon=epsilon, dtype=np.float64)
+        # get sample coefficients
+        x = self.create_images(L, 1)
+        # hold input test data constant (would depend on epsilon parameter)
+        evaluate_t = basis.evaluate(basis.evaluate_t(x))
+        expand = basis.evaluate(basis.expand(evaluate_t))
+
+        relerr = self.relerr(evaluate_t.asnumpy(), expand.asnumpy())
```

Comment:
If `expand` is more accurate I think you should switch the order here: `relerr = self.relerr(expand.asnumpy(), evaluate_t.asnumpy())`. See my comment below where you define `relerr`.

Reply:
Good catch thanks. Changed